[AIRFLOW-2965] CLI tool to show the next execution datetime

XD-DENG opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-2965
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

This PR is to address the request proposed by JIRA user Jack , in JIRA ticket [2965](https://issues.apache.org/jira/browse/AIRFLOW-2965).
@r39132 suggested a PR in this ticket.

This new CLI tool can help show the next execution datetime, given DAG id. Under the hood, its checking the following schedule of the latest execution datetime.

Let's say I have DAG `basic_dag` whose schedule_interval is `*/2 * * * *`. The latest execution_date is 2018-08-31 10:36:00 (start time is 2018-08-31 10:38:00).

If I run this new CLI tool `airflow next_execution basic_dag`, the result would be 
<img width="855" alt="screen shot 2018-09-02 at 10 29 05 pm" src="https://user-images.githubusercontent.com/11539188/44957068-ef54b680-aeff-11e8-831d-9319b85135d2.png">

After I change the schedule_interval to `0 23 * * *` and run the same command again, the result is 
<img width="868" alt="screen shot 2018-09-02 at 10 30 02 pm" src="https://user-images.githubusercontent.com/11539188/44957070-f8de1e80-aeff-11e8-9785-87ffaf2e74ce.png">
**(actual start time would be 2018-09-01 23:00:00)**

Again, we change the scheduler_interval to `0 3 * * *`, the result is 
<img width="876" alt="screen shot 2018-09-02 at 10 30 42 pm" src="https://user-images.githubusercontent.com/11539188/44957071-faa7e200-aeff-11e8-83a1-f25a008394ac.png">
**(actual start time would be 2018-09-02 03:00:00)**

Running with `-h` would return the usage and arguments needed.
<img width="662" alt="screen shot 2018-09-02 at 10 36 14 pm" src="https://user-images.githubusercontent.com/11539188/44957110-a2251480-af00-11e8-97df-cd8077c6ba76.png">




### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

I believe the documentation generation of CLI is self-contained in script `airflow/bin/cli.py`.

### Code Quality

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

r39132 commented on airflow/bin/cli.py:
```diff
@@ -551,6 +551,17 @@ def dag_state(args):
     print(dr[0].state if len(dr) > 0 else None)
 
 
+@cli_utils.action_logging
```
@XD-DENG I just tested this with some of the example dags in https://github.com/apache/incubator-airflow/tree/master/airflow/example_dags. Can you test your code with different schedule types including `@once`, `daily/weekly`, `timedelta(hours=1)`, etc... in addition to the case you provide which is cron expressions. Also, can you add tests for these?

```
(venv) sianand@LM-SJN-21002367:~/Projects/airflow_incubator $ airflow next_execution latest_only
[2018-09-04 10:52:19,613] {__init__.py:51} INFO - Using executor SequentialExecutor
/Users/sianand/miniconda3/lib/python3.6/site-packages/apache_airflow-2.0.0.dev0+incubating-py3.6.egg/airflow/bin/cli.py:1724: DeprecationWarning: The celeryd_concurrency option in [celery] has been renamed to worker_concurrency - the old setting has been used, but please update your config.
  default=conf.get('celery', 'worker_concurrency')),
[2018-09-04 10:52:19,822] {models.py:260} INFO - Filling up the DagBag from /Users/sianand/Projects/airflow_incubator/dags
[2018-09-04 10:52:19,882] {example_kubernetes_operator.py:55} WARNING - Could not import KubernetesPodOperator: No module named 'kubernetes'
[2018-09-04 10:52:19,882] {example_kubernetes_operator.py:56} WARNING - Install kubernetes dependencies with:     pip install apache-airflow[kubernetes]
Traceback (most recent call last):
  File "/Users/sianand/miniconda3/bin/airflow", line 4, in <module>
    __import__('pkg_resources').run_script('apache-airflow==2.0.0.dev0+incubating', 'airflow')
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/pkg_resources/__init__.py", line 654, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/pkg_resources/__init__.py", line 1434, in run_script
    exec(code, namespace, namespace)
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/apache_airflow-2.0.0.dev0+incubating-py3.6.egg/EGG-INFO/scripts/airflow", line 32, in <module>
    args.func(args)
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/apache_airflow-2.0.0.dev0+incubating-py3.6.egg/airflow/utils/cli.py", line 74, in wrapper
    return f(*args, **kwargs)
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/apache_airflow-2.0.0.dev0+incubating-py3.6.egg/airflow/bin/cli.py", line 562, in next_execution
    print(dag.following_schedule(dag.latest_execution_date))
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/apache_airflow-2.0.0.dev0+incubating-py3.6.egg/airflow/models.py", line 3371, in following_schedule
    return dttm + self._schedule_interval
TypeError: unsupported operand type(s) for +: 'NoneType' and 'datetime.timedelta'


```

r39132 reviewed (changes requested):
@XD-DENG I'm adding some more details here:

I'm seeing multiple different issues during testing, so I'll document them here:
* example_bash_operators (`one of the example dags`)
Before the first run of a DAG, the exception below is raised:
```
(venv) sianand@LM-SJN-21002367:~/Projects/airflow_incubator $ airflow next_execution  example_bash_operator 
[2018-09-04 13:44:07,847] {__init__.py:51} INFO - Using executor SequentialExecutor
/Users/sianand/miniconda3/lib/python3.6/site-packages/apache_airflow-2.0.0.dev0+incubating-py3.6.egg/airflow/bin/cli.py:1724: DeprecationWarning: The celeryd_concurrency option in [celery] has been renamed to worker_concurrency - the old setting has been used, but please update your config.
  default=conf.get('celery', 'worker_concurrency')),
[2018-09-04 13:44:07,988] {models.py:260} INFO - Filling up the DagBag from /Users/sianand/Projects/airflow_incubator/dags
[2018-09-04 13:44:08,032] {example_kubernetes_operator.py:55} WARNING - Could not import KubernetesPodOperator: No module named 'kubernetes'
[2018-09-04 13:44:08,032] {example_kubernetes_operator.py:56} WARNING - Install kubernetes dependencies with:     pip install apache-airflow[kubernetes]
Traceback (most recent call last):
  File "/Users/sianand/miniconda3/bin/airflow", line 4, in <module>
    __import__('pkg_resources').run_script('apache-airflow==2.0.0.dev0+incubating', 'airflow')
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/pkg_resources/__init__.py", line 654, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/pkg_resources/__init__.py", line 1434, in run_script
    exec(code, namespace, namespace)
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/apache_airflow-2.0.0.dev0+incubating-py3.6.egg/EGG-INFO/scripts/airflow", line 32, in <module>
    args.func(args)
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/apache_airflow-2.0.0.dev0+incubating-py3.6.egg/airflow/utils/cli.py", line 74, in wrapper
    return f(*args, **kwargs)
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/apache_airflow-2.0.0.dev0+incubating-py3.6.egg/airflow/bin/cli.py", line 562, in next_execution
    print(dag.following_schedule(dag.latest_execution_date))
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/apache_airflow-2.0.0.dev0+incubating-py3.6.egg/airflow/models.py", line 3366, in following_schedule
    dttm = timezone.make_naive(dttm, self.timezone)
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/apache_airflow-2.0.0.dev0+incubating-py3.6.egg/airflow/utils/timezone.py", line 138, in make_naive
    if is_naive(value):
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/apache_airflow-2.0.0.dev0+incubating-py3.6.egg/airflow/utils/timezone.py", line 49, in is_naive
    return value.utcoffset() is None
AttributeError: 'NoneType' object has no attribute 'utcoffset'
```

After it has one run, then it works.
```
(venv) sianand@LM-SJN-21002367:~/Projects/airflow_incubator $ airflow next_execution  example_bash_operator 
[2018-09-04 14:28:16,680] {__init__.py:51} INFO - Using executor SequentialExecutor
[2018-09-04 14:28:16,885] {models.py:260} INFO - Filling up the DagBag from /Users/sianand/Projects/airflow_incubator/dags
2018-09-04 00:00:00+00:00
```
Hence, I think the command only makes sense when the start date is know or at least one execution is known.

* latest_only (`one of the example dags`)
Before the first run of a DAG, the exception below is raised:

```
(venv) sianand@LM-SJN-21002367:~/Projects/airflow_incubator $ airflow next_execution latest_only
[2018-09-04 14:30:20,321] {__init__.py:51} INFO - Using executor SequentialExecutor
[2018-09-04 14:30:20,517] {models.py:260} INFO - Filling up the DagBag from /Users/sianand/Projects/airflow_incubator/dags
Traceback (most recent call last):
  File "/Users/sianand/miniconda3/bin/airflow", line 4, in <module>
    __import__('pkg_resources').run_script('apache-airflow==2.0.0.dev0+incubating', 'airflow')
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/pkg_resources/__init__.py", line 654, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/pkg_resources/__init__.py", line 1434, in run_script
    exec(code, namespace, namespace)
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/apache_airflow-2.0.0.dev0+incubating-py3.6.egg/EGG-INFO/scripts/airflow", line 32, in <module>
    args.func(args)
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/apache_airflow-2.0.0.dev0+incubating-py3.6.egg/airflow/utils/cli.py", line 74, in wrapper
    return f(*args, **kwargs)
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/apache_airflow-2.0.0.dev0+incubating-py3.6.egg/airflow/bin/cli.py", line 562, in next_execution
    print(dag.following_schedule(dag.latest_execution_date))
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/apache_airflow-2.0.0.dev0+incubating-py3.6.egg/airflow/models.py", line 3371, in following_schedule
    return dttm + self._schedule_interval
TypeError: unsupported operand type(s) for +: 'NoneType' and 'datetime.timedelta'(venv) sianand@LM-SJN-21002367:~/Projects/airflow_incubator $ airflow next_execution latest_only
[2018-09-04 14:30:20,321] {__init__.py:51} INFO - Using executor SequentialExecutor
[2018-09-04 14:30:20,517] {models.py:260} INFO - Filling up the DagBag from /Users/sianand/Projects/airflow_incubator/dags
Traceback (most recent call last):
  File "/Users/sianand/miniconda3/bin/airflow", line 4, in <module>
    __import__('pkg_resources').run_script('apache-airflow==2.0.0.dev0+incubating', 'airflow')
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/pkg_resources/__init__.py", line 654, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/pkg_resources/__init__.py", line 1434, in run_script
    exec(code, namespace, namespace)
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/apache_airflow-2.0.0.dev0+incubating-py3.6.egg/EGG-INFO/scripts/airflow", line 32, in <module>
    args.func(args)
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/apache_airflow-2.0.0.dev0+incubating-py3.6.egg/airflow/utils/cli.py", line 74, in wrapper
    return f(*args, **kwargs)
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/apache_airflow-2.0.0.dev0+incubating-py3.6.egg/airflow/bin/cli.py", line 562, in next_execution
    print(dag.following_schedule(dag.latest_execution_date))
  File "/Users/sianand/miniconda3/lib/python3.6/site-packages/apache_airflow-2.0.0.dev0+incubating-py3.6.egg/airflow/models.py", line 3371, in following_schedule
    return dttm + self._schedule_interval
TypeError: unsupported operand type(s) for +: 'NoneType' and 'datetime.timedelta'
```

After the first run, it works:
```
(venv) sianand@LM-SJN-21002367:~/Projects/airflow_incubator $ airflow next_execution latest_only
[2018-09-04 14:33:06,591] {__init__.py:51} INFO - Using executor SequentialExecutor
[2018-09-04 14:33:06,758] {models.py:260} INFO - Filling up the DagBag from /Users/sianand/Projects/airflow_incubator/dags
2018-09-04 20:00:00+00:00
```

These are both examples of what happens when the first run time is missing, but they use different scheduling schemes (e.g. cron, timedelta, `@once`).

Please cover this case and add tests for these different scheduling schemes.

XD-DENG committed:
[AIRFLOW-2965] CLI tool to show the next execution datetime

Cover different cases

- schedule_interval is "@once" or None, then following_schedule
  method would always return None
- If dag is paused, print reminder
- If latest_execution_date is not found, print warning saying
  not applicable.
